glommio/src/io/bulk_io.rs

```diff
@@ -49,3 +49,52 @@ impl<U: Copy + Unpin> Stream for OrderedBulkIo<U> {
         (self.iovs.len(), Some(self.iovs.len()))
     }
 }
+
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct ReadManyArgs {
+    pub(crate) user_read: (u64, usize),
+    pub(crate) system_read: (u64, usize),
+}
+
+#[derive(Debug)]
+pub struct ReadManyResult {
+    pub(crate) inner: OrderedBulkIo<ReadManyArgs>,
+    pub(crate) current_result: ReadResult,
+}
+
+impl Stream for ReadManyResult {
+    type Item = super::Result<(u64, ReadResult)>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let next = match Pin::new(&mut self.inner).poll_next(cx) {
+            Poll::Ready(next) => next,
+            Poll::Pending => return Poll::Pending,
+        };
+
+        match next {
+            None => Poll::Ready(None),
+            Some((source, args)) => {
+                if let Some(mut source) = source {
+                    let read_size =
+                        enhanced_try!(source.take_result().unwrap(), "Reading", self.inner.file)?;
+                    let mut buffer = source.extract_dma_buffer();
+                    buffer.trim_to_size(std::cmp::min(read_size, args.system_read.1));
```

Comment:
why do we need to trim, if this buffer won't ever be visible to the user anyway ?

Reply:
Correct; I was trimming twice, will remove